schoenrock-rossiter_tom_ndvi.py

```diff
@@ -0,0 +1,656 @@
+
+    return band
+
+# function to apply cloud mask and caluclate ndvi (mean)
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   masked_values):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        a list containing the xarray objects for landsat  bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using
+        rasterio mask().
+    pixel_qa: xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been
+        turned into a mask (0s and 1s)
+    masked_values: list
+        A list of cloud mask values
+
+
+    Returns
+    -----------
+    ndvi_crop : Xarray Dataset
+        a cropped and masked xarray object containing NDVI values
+    """
+
+    crop_json = crop_bound.geometry
+
+    # Clip pixel qa cloud mask layer
+    cl_mask_crop = pixel_qa.rio.clip(crop_json)
+
+    # Calculate NDVI
+    ndvi_xr = (all_bands[1]-all_bands[0]) / (all_bands[1]+all_bands[0])
+    # Clip NDVI layer
+    ndvi_crop = ndvi_xr.rio.clip(crop_json)
+
+    # Apply cloud mask to NDVI
+    ndvi_crop = ndvi_crop.where(~pixel_qa.isin(masked_values))
+
+    
+    return ndvi_crop.mean().values
+    
+#third function containing the above two
+def output_ndvi(scene_path, crop_extent):
+    """Function that combines the above two functions to open, stack, crop
+    and mask Landsat scenes.
+
+    Parameters
+    -----------
+    scene_path : string
+        A path to the array to be opened
+    crop_extent: vector
+        A path to the vector file to use as crop boundary
+    
+    Returns
+    -----------
+    ndvi_crop : ndarray Dataset
+        a cropped and masked ndarray object containing mean NDVI values
+    """
+    
+    band_paths = glob(os.path.join(scene_path, "*band*[4-5].tif"))
+    pixel_qa_path = glob(os.path.join(scene_path, "*qa*"))
+    landsat_qa = rxr.open_rasterio(pixel_qa_path[0], masked = True).squeeze()
+    
+    high_cloud_confidence = em.pixel_flags["pixel_qa"]["L8"]    ["High Cloud Confidence"]
+    cloud = em.pixel_flags["pixel_qa"]["L8"]["Cloud"]
+    cloud_shadow = em.pixel_flags["pixel_qa"]["L8"]["Cloud Shadow"]
+
+    all_masked_values = cloud_shadow + cloud + high_cloud_confidence
+    
+    #run clean_bands function
+    bands = []
+    for band_path in band_paths:
+        cleaned_band = open_clean_bands(band_path = band_path,
+                                    crop_extent = crop_extent,
+                                    valid_range = (0, 10000))
+        bands.append(cleaned_band)
+        
+    #run mask_crop function
+    ndvi = mask_crop_ndvi(all_bands = bands,
+                        crop_bound = crop_extent,
+                        pixel_qa = landsat_qa,
+                        masked_values = all_masked_values)
+    
+    #print(ndvi)
+    return ndvi
+
+ndvi_list = []
+
```

Comment:
I would find it helpful if there were comments explaining what these two nested for loops are designed for.

Reply:
I have addressed this with further comments